src/python/pants/backend/java/util_rules.py

```diff
@@ -3,79 +3,119 @@
+            f"{self.java_home}/bin/java",
+            "-cp",
+            ":".join([self.nailgun_jar, *classpath_entries]),
+        )
 
 
 @rule
-async def setup_jdk(coursier: Coursier, javac: JavacSubsystem) -> JdkSetup:
+async def setup_jdk(coursier: Coursier, javac: JavacSubsystem, bash: BashBinary) -> JdkSetup:
+    nailgun = await Get(
+        ResolvedClasspathEntry,
+        CoursierLockfileEntry(
+            coord=Coordinate.from_coord_str("com.martiansoftware:nailgun-server:0.9.1"),
+            file_name="nailgun-server-0.9.1.jar",
+            direct_dependencies=Coordinates(),
+            dependencies=Coordinates(),
+            file_digest=FileDigest(
+                fingerprint="4518faa6bf4bd26fccdc4d85e1625dc679381a08d56872d8ad12151dda9cef25",
+                serialized_bytes_length=32927,
+            ),
+        ),
+    )
+
     if javac.options.jdk == "system":
         coursier_jdk_option = "--system-jvm"
     else:
         coursier_jdk_option = f"--jvm={javac.options.jdk}"
+    java_home_command = f"{coursier.coursier.exe} java-home {coursier_jdk_option}"
 
-    java_home_result = await Get(
+    java_version_result = await Get(
         FallibleProcessResult,
         Process(
             argv=(
-                coursier.coursier.exe,
-                "java-home",
-                coursier_jdk_option,
+                bash.path,
+                "-c",
+                f"$({java_home_command})/bin/java -version",
             ),
             input_digest=coursier.digest,
             description=f"Ensure download of JDK {coursier_jdk_option}.",
             cache_scope=ProcessCacheScope.PER_RESTART_SUCCESSFUL,
         ),
     )
 
-    if java_home_result.exit_code != 0:
+    if java_version_result.exit_code != 0:
         raise ValueError(
-            f"Failed to determine JAVA_HOME for JDK {javac.options.jdk}: {java_home_result.stderr.decode('utf-8')}"
+            f"Failed to locate Java for JDK `{javac.options.jdk}`:\n"
+            f"{java_version_result.stderr.decode('utf-8')}"
         )
 
-    java_home = java_home_result.stdout.decode("utf-8").strip()
+    java_version = java_version_result.stdout.decode("utf-8").strip()
 
-    version_result = await Get(
-        ProcessResult,
-        Process(
-            argv=(
-                f"{java_home}/bin/java",
-                "-version",
-            ),
-            description=f"Extract version from JDK {coursier_jdk_option}.",
-            cache_scope=ProcessCacheScope.PER_RESTART_SUCCESSFUL,
-        ),
-    )
+    # TODO: Locate `ln`.
+    jdk_preparation_script = textwrap.dedent(
+        f"""\
+        # pants javac script using Coursier {coursier_jdk_option}. `java -version`: {java_version}"
+        set -eu
 
-    all_output = "\n".join(
-        [
-            version_result.stderr.decode("utf-8"),
-            version_result.stdout.decode("utf-8"),
-        ]
+        /bin/ln -s "$({java_home_command})" "{JdkSetup.java_home}"
```

Comment:
Why is this symlinking JAVA_HOME within the input root?

Reply:
To allow callers to consume it as a relative path: https://github.com/pantsbuild/pants/pull/12982/files#diff-3dd6edcc8f9a9ef9521eb1b9cec604222276eefafa48d966fc5378e4dce20148R246-R247